src/gen_rules.ml

```diff
@@ -689,20 +689,44 @@ Add it to your jbuild file to remove this warning.
       guess_modules ~dir:src_dir
         ~files:(Lazy.force files))
     in
-    List.filter_map stanzas ~f:(fun stanza ->
-      let dir = ctx_dir in
-      match (stanza : Stanza.t) with
-      | Library lib  ->
-        Some (library_rules lib ~dir
-                ~all_modules:(Lazy.force all_modules) ~files:(Lazy.force files)
-                ~scope)
-      | Executables  exes ->
-        Some (executables_rules exes ~dir ~all_modules:(Lazy.force all_modules)
-                ~scope)
-      | _ -> None)
-    |> Merlin.add_rules sctx ~dir:ctx_dir
-
-  let () = List.iter (SC.stanzas sctx) ~f:rules
+    let merlin =
+      List.filter_map stanzas ~f:(fun stanza ->
+        let dir = ctx_dir in
+        match (stanza : Stanza.t) with
+        | Library lib ->
+          Some (library_rules lib ~dir
+                  ~all_modules:(Lazy.force all_modules) ~files:(Lazy.force files)
+                  ~scope)
+        | Executables exes ->
+          Some (executables_rules exes ~dir ~all_modules:(Lazy.force all_modules)
+                  ~scope)
+        | _ -> None)
+      |> Merlin.merge_all in
+    Option.iter merlin ~f:(Merlin.add_rules sctx ~dir:ctx_dir);
+    Option.map merlin ~f:(fun m -> m.Merlin.requires)
+
+  let utop_rules lib_requires =
+    let ctx_dir = (SC.context sctx).build_dir in
+    let scope = Scope.empty in
```

Comment:
Oh, so basically what I had before but without the user providing a library name as a command line argument. But why have the user cd around to different directories? Isn't it more convenient to pass the directory name as an argument?

What's the issue of having per library top levels though? I don't see why per directory is better off the top of my head.